File: cplus/leetcode/202606/algoiv08.11.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int waysToChange(int n) {
        // 完全背包
        const int mod = 1e9 + 7;
        vector<int> dp(n+1), coins = {1, 5, 10, 25};
        dp[0] = 1;
        for (auto coin: coins) for (int w=coin; w<=n; w++) {
            dp[w] = (dp[w] + dp[w-coin]) % mod;
        }

        return dp[n];
    }
};

File: python/algo/202606/algoiv08.11.py
class Solution:
    def waysToChange(self, n: int) -> int:
        mod = 10 ** 9 + 7
        coins = [1, 5, 10, 25]
        # 完全背包
        dp = [0] * (n+1)
        dp[0] = 1
        for coin in coins:
            for w in range(coin, n+1):
                dp[w] = (dp[w] + dp[w-coin]) % mod

        return dp[n]